draft-pda-core.md

```diff
@@ -291,11 +293,427 @@ Each PA task in this document is divided into three sub-protocols as follows.
+support for multiple cipher suites.]
+
+We call this the helper's *key configation*. The key configuration is used to
+set up a base-mode HPKE context to use to derive symmetric keys for protecting
+the shares sent to the helper. The *config id*, `HpkeConfig.id`, is forwarded
+by the client to the helper, who uses this value to decide if it knows how to
+decrypt a share it receives.
+
+## Pre-conditions
+
+We assume the following conditions hold before the client begins uploading its
+data:
+1. The client, aggregators, and collector are configured with a specific PA task.
+1. The client knows the URL of the leader endpoint, e.g., `example.com/metrics`.
+   We write this URL as `[leader]` below. (We write `[helper]` for a helper's
+   URL.)
+1. The client and leader can establish a leader-authenticated secure channel.
+1. The leader and each helper can establish a leader-authenticated secure
+   channel.
+1. Each helper has chosen an HPKE key pair.
+1. The aggregators agree on a set of PA tasks, as well as the PA protocol and
+   parameters used for each task.
+
+[TODO: It would be clearer to include a "pre-conditions" section prior to each
+"phase" of the protocol.]
+
+## Upload {#pa-upload}
+
+[TODO: Add an illustration of this sub-protocol.]
+
+Uploading a report involves two requests to the leader. In the  *upload start
+request*, the client discovers the protocol-specific parameters it needs to
+generate the report, as well as the endpoint URL of each helper. In the *upload
+finish request*, it uploads its report to the leader.
+
+[NOTE: @acmiyaguchi pointed out that the use of an anonymizing proxy for
+uploading shares might be easier to implement if the "upload" phase involved a
+single HTTP request. However, OHTTP
+(https//www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html) allows
+clients to make multiple requests through a proxy, so these kinds of use cases
+should work.]
+
+[NOTE: @cjpatton: Breaking the upload phase into two requests is useful for
+supporting Prio-like proof systems in which the leader sends the client a
+"challenge" it uses to generate the proof. (See the SIMD construction of
+{{BBCp19}}, Section 5.2. Here, the "challenge" is a randomly generated field
+element.) Because the client/leader are using TLS, another way to implement this
+would be to export a secret from the key schedule.]
+
+### Upload Start
+
+The client sends a POST request to `[leader]/upload_start` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+} PAUploadStartReq;
+~~~
+
+The `task` field corresponds to the PA task for which a report will be
+generated.
+
+The leader responds to well-formed requests to `[leader]/upload_start` with
+status 200 and the following message:
+
+~~~
+struct {
+  PAParam param;
+  Url helper_urls<1..2^16-1>;
+  select (PAUploadStartResp.param.proto) {
+    case prio: PrioUploadStartResp;
+    case hits: HitsUploadStartResp;
+  }
+} PAUploadStartResp;
```

Comment:
I don't quite understand the underlying assumptions here. If the client wants to upload some data, it's not going to be confused about whether the task involves heavy hitters or prio. What happens if it thinks it's doing prio and the leader responds with heavy hitterS?

Reply:
This is to accommodate different parameters for the actual upload that are protocol specific. For example, if Hits or some future protocol requires per-upload data, that'd be in the HitsUploadStartResp.

Reply:
Sure, but why is PAParam here? Almost nothing there can change.

Reply:
I guess it depends on how things are configured? The assumption now is that clients start with a Task and then learn the Params for that task during Upload, but if we assume that Params are also distributed out-of-band, then we can drop it here.